Make time picker accessible

yjbanov opened the pull request:
Fixes https://github.com/flutter/flutter/issues/1667 and maybe https://github.com/flutter/flutter/issues/1666 (it seems adding custom nodes isn't that hard)

/cc @goderbauer @Hixie

yjbanov committed:
make time picker accessible

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -762,6 +767,78 @@ class _DialPainter extends CustomPainter {
+                ..textDirection = textDirection
+                ..addAction(SemanticsAction.tap, () {
+                  // The action is sent to the gesture detector surrounding the
+                  // custom painter.
```
is that supposed to happen?

cc @goderbauer

yjbanov replied:
I'll wait for @goderbauer's OK before submitting, but I checked a couple of things:

- instrumented `Window.onPointerDataPacket` and made sure the OS sends taps to the selected rectangle irrespective of where you double-tap, which tells me that the OS gets the semantic nodes created by the time picker.
- checked with other widgets (e.g. `CupertinoSwitch`) and found that they behave the same way, i.e. the closure passed to `addAction` is not invoked, and instead is sent to the gesture recognizer.

goderbauer replied:
This sounds like a bug. I am investigating...

goderbauer replied:
Turns out: The observed behavior is not a bug, but a feature in Android prior to Oreo, see https://github.com/flutter/flutter/issues/13219 for discussion.

Having said that, to make the widget also accessible for newer Android versions (Oreo and newer) as well as iOS, this handler needs to implement the actual logic for `SemanticsAction.tap`.

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -762,6 +767,78 @@ class _DialPainter extends CustomPainter {
+                })
+          )
+          ..tags = new Set<SemanticsTag>.from(const <SemanticsTag>[
+            const SemanticsTag('dial-label'),
```
add a comment saying this is only used by tests.

yjbanov replied:
Done.

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -762,6 +767,78 @@ class _DialPainter extends CustomPainter {
+            const SemanticsTag('dial-label'),
+          ]);
+
+        nodes.add(node);
```
I assume the nodes are traversed in geometrical order, not time order?

cc @gspencergoog who is working on making that fixable -- fixing the order here would be a good proof of that patch.

yjbanov replied:
Aye! @gspencergoog, LMK if you need help.

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1069,12 +1158,56 @@ class _DialState extends State<_Dial> with SingleTickerProviderStateMixin {
+  void describeSemanticsConfiguration(SemanticsConfiguration config) {
+    super.describeSemanticsConfiguration(config);
+    config.isSemanticBoundary = true;
+    config.textDirection = TextDirection.ltr;
```
why ltr? aren't the arabic labels rtl?

yjbanov replied:
Good catch. This is left-over from my experiments. Removed.

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1069,12 +1158,56 @@ class _DialState extends State<_Dial> with SingleTickerProviderStateMixin {
+  }
+}
+
+class _RenderDialPaint extends RenderCustomPaint {
```
might be worth folding these improvements into RenderCustomPaint and CustomPaint. That would actually fix https://github.com/flutter/flutter/issues/1666.

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1105,9 +1238,17 @@ class _TimePickerDialogState extends State<_TimePickerDialog> {
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    localizations = MaterialLocalizations.of(context);
+    _announceToAccessibility(localizations.timePickerHourModeAnnouncement);
```
This means every time any inherited widget that this widget depends on changes, we'll announce this. That doesn't seem right. What if it's, e.g., the TickerMode? Or the Theme, while the theme is animating?

yjbanov replied:
Done.

Hixie commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1266,8 +1414,23 @@ Future<TimeOfDay> showTimePicker({
+    'data': <String, dynamic>{
+      'message': message,
+    },
+  });
```
i thought we had an API for this.

cc @amirh, who I think was working on that.

amirh replied:
Yes we have [SemanticsService.announce](https://master-docs-flutter-io.firebaseapp.com/flutter/semantics/SemanticsService/announce.html)

yjbanov replied:
Ooh! Thanks for adding this. Done.

Hixie commented on packages/flutter/test/rendering/mock_canvas.dart:
```diff
@@ -282,8 +282,15 @@ abstract class PaintPattern {
   /// [Canvas.drawParagraph] call's argument, and any mismatches result in failure.
   ///
+  /// The `offset` argument can be either an [Offset] or a [Matcher]. If it is
+  /// and [Offset] then the actual value much match the expected offset
```
and->an

yjbanov replied:
Done. Also "much match"->"must match"

Hixie commented:
I think we should also make the hours and minutes in the header support being incremented and decremented using the relevant semantics actions (like a slider).

yjbanov commented:
Comments from @goderbauer:

- [ ] fix https://github.com/flutter/flutter/issues/1666#issuecomment-347371383 first, apply here, then fix the rest:
- [ ] if a value is already selected, the announcement should read "Selected {VALUE}" (everywhere, header values, dial, am/pm)
- [ ] actions on semantics nodes must be position independent, because iOS does not report semantics action coordinates
- [ ] fix AM/PM
- [ ] talk to @gspencergoog to make sure this works with the upcoming traversal order infra

goderbauer commented:
One more comment to add to the list above: The GestureDetector around the entire CustomPaint area should be excluded from Semantics (it is currently focusable on iOS at least without providing any value).

yjbanov committed:
use new CustomPaint a11y API

yjbanov committed:
flutter_localizations tests; use bigger distance delta

yjbanov committed:
fix am/pm control; selected values

yjbanov commented:
Ok, moved to the new `CustomPainterSemantics` API (https://github.com/flutter/flutter/pull/13152/commits/732d94c6be6cb90e3a6bbdfbd1cb62a20b179c87) and implemented selected values along with a new am/pm control (https://github.com/flutter/flutter/pull/13152/commits/f4a7cc7ced87c03baa9c9c5219e6fb3a641ed7f9).

PTAL.

goderbauer commented on packages/flutter/lib/src/rendering/object.dart:
````diff
@@ -2149,6 +2151,7 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   /// }
   /// ```
   @protected
+  @mustCallSuper
````
not sure about this one. Why should I not be able to create a subclass that throws away all semantic information of the super class?

yjbanov replied:
Removed.

goderbauer commented on packages/flutter_localizations/lib/src/l10n/localizations.dart:
```diff
@@ -77,6 +79,8 @@ const Map<String, Map<String, String>> localizations = const <String, Map<String
     'viewLicensesButtonLabel': r'LIZENZEN ANZEIGEN',
     'anteMeridiemAbbreviation': r'VORM.',
     'postMeridiemAbbreviation': r'NACHM.',
+    'timePickerHourModeAnnouncement': r'Wählen Sie Stunden',
```
Stunde auswählen

yjbanov replied:
Done.

goderbauer commented on packages/flutter_localizations/lib/src/l10n/localizations.dart:
```diff
@@ -77,6 +79,8 @@ const Map<String, Map<String, String>> localizations = const <String, Map<String
     'anteMeridiemAbbreviation': r'VORM.',
     'postMeridiemAbbreviation': r'NACHM.',
+    'timePickerHourModeAnnouncement': r'Wählen Sie Stunden',
+    'timePickerMinuteModeAnnouncement': r'Wählen Sie Minuten',
```
Minute auswählen

yjbanov replied:
Done.

goderbauer commented on packages/flutter_localizations/lib/src/l10n/localizations.dart:
```diff
@@ -140,6 +144,8 @@ const Map<String, Map<String, String>> localizations = const <String, Map<String
     'anteMeridiemAbbreviation': r'AM',
     'postMeridiemAbbreviation': r'PM',
+    'timePickerHourModeAnnouncement': r'Select hours.',
+    'timePickerMinuteModeAnnouncement': r'Select minutes.',
```
Why do these two strings end with a full stop? None of the other english translations seem to have it. If they should, should the other languages include it as well?

yjbanov replied:
Removed.

goderbauer commented on packages/flutter_localizations/lib/src/l10n/material_de.arb:
```diff
@@ -28,5 +28,7 @@
   "anteMeridiemAbbreviation": "VORM.",
-  "postMeridiemAbbreviation": "NACHM."
+  "postMeridiemAbbreviation": "NACHM.",
+  "timePickerHourModeAnnouncement": "Wählen Sie Stunden",
```
Same as above

yjbanov replied:
Done.

goderbauer commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -235,13 +275,18 @@ class _HourControl extends StatelessWidget {
-      ), style: hourStyle),
+      child: new Semantics(
+        selected: fragmentContext.mode == _TimePickerMode.hour,
+        hint: localizations.timePickerHourModeAnnouncement,
```
The hint usually tells users what interactions have what effect. So this one should probably say "double tap to select hours" instead of just "select hours".

goderbauer replied:
If it's not reflected in the docs for hint, we should add it there.

yjbanov replied:
Hmm, this only announces that the time picker changed mode to either hour or minute. There is no one action that will allow you to select hours, as there are many controls on the screen. This is consistent with what Android does. It just says "select hours"/"select minutes". Although perhaps in a future iteration of the time picker we could improve the situation.

goderbauer commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -281,7 +328,11 @@ class _MinuteControl extends StatelessWidget {
-      child: new Text(localizations.formatMinute(fragmentContext.selectedTime), style: minuteStyle),
+      child: new Semantics(
+        selected: fragmentContext.mode == _TimePickerMode.minute,
+        hint: localizations.timePickerMinuteModeAnnouncement,
```
same

yjbanov replied:
ditto

goderbauer reviewed (commented):
Can you file issues about any remaining issues with the time picker?

cbracken commented on packages/flutter_localizations/lib/src/l10n/material_ja.arb:
```diff
@@ -27,5 +27,7 @@
   "anteMeridiemAbbreviation": "AM",
-  "postMeridiemAbbreviation": "PM"
+  "postMeridiemAbbreviation": "PM",
+  "timePickerHourModeAnnouncement": "時間の選択",
```
時の選択

yjbanov replied:
Done.

cbracken commented on packages/flutter_localizations/lib/src/l10n/localizations.dart:
```diff
@@ -588,6 +604,8 @@ const Map<String, Map<String, String>> localizations = const <String, Map<String
     'viewLicensesButtonLabel': r'ライセンスを表示',
     'anteMeridiemAbbreviation': r'AM',
     'postMeridiemAbbreviation': r'PM',
+    'timePickerHourModeAnnouncement': r'時間の選択',
```
時を選択

yjbanov replied:
Done.

cbracken commented on packages/flutter_localizations/lib/src/l10n/material_fr.arb:
```diff
@@ -28,5 +28,7 @@
-  "postMeridiemAbbreviation": "PM"
+  "postMeridiemAbbreviation": "PM",
+  "timePickerHourModeAnnouncement": "Sélectionnez les heures",
+  "timePickerMinuteModeAnnouncement": "Sélectionnez les minutes"
```
French translations lgtm.

yjbanov replied:
\o/

yjbanov committed:
fix translations; remove @mustCallSuper in describeSemanticsConfiguration

yjbanov commented:
> Can you file issues about any remaining issues with the time picker?

https://github.com/flutter/flutter/issues/13400
https://github.com/flutter/flutter/issues/13453

goderbauer commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -183,9 +184,26 @@ class _DayPeriodControl extends StatelessWidget {
+    if (fragmentContext.selectedTime.period == DayPeriod.pm) {
+      return;
+    }
+    _announceToAccessibility(context, MaterialLocalizations.of(context).postMeridiemAbbreviation);
```
Should we exclude these announcements on iOS since they are not necessary?

yjbanov replied:
Done.

yjbanov replied:
https://github.com/flutter/flutter/pull/13152/commits/00f7af9b122d09c7a47e49ab08557e083527189e

goderbauer reviewed (approved):
LGTM

yjbanov committed:
exclude AM/PM announcement from iOS as on iOS the label is read back automatically
